src/duplicates_impl.rs

```diff
@@ -0,0 +1,197 @@
+use std::hash::Hash;
+use std::collections::HashMap;
+
+mod private {
+    use std::fmt;
+    use std::hash::Hash;
+    use std::collections::HashMap;
+
+    #[derive(Clone)]
+    #[must_use = "iterator adaptors are lazy and do nothing unless consumed"]
+    pub struct DuplicatesBy<I: Iterator, Key, F> {
+        pub(crate) iter: I,
+        pub(crate) used: HashMap<Key, bool>,
+        pub(crate) pending: usize,
+        pub(crate) key_method: F,
+    }
+
+    impl<I, V, F> fmt::Debug for DuplicatesBy<I, V, F>
+        where I: Iterator + fmt::Debug,
+              V: fmt::Debug + Hash + Eq,
+    {
+        debug_fmt_fields!(DuplicatesBy, iter, used);
+    }
+
+    impl<I, Key, F> DuplicatesBy<I, Key, F>
+        where I: Iterator,
+              Key: Eq + Hash,
+              F: KeyMethod<Key, I::Item>,
+    {
+        /// Takes an item and returns it back to the caller if it's the second time we see it.
+        /// Otherwise the item is consumed and None is returned
+        fn filter(&mut self, item: I::Item) -> Option<I::Item> {
+            let kv = self.key_method.make(item);
+            match self.used.get_mut(kv.key_ref()) {
+                None => {
+                    self.used.insert(kv.key(), false);
+                    self.pending += 1;
+                    None
+                },
+                Some(true) => None,
+                Some(produced) => {
+                    *produced = true;
+                    self.pending -= 1;
+                    Some(kv.value())
+                }
+            }
+        }
+    }
+
+    impl<I, Key, F> Iterator for DuplicatesBy<I, Key, F>
+        where I: Iterator,
+              Key: Eq + Hash,
+              F: KeyMethod<Key, I::Item>,
+    {
+        type Item = I::Item;
+
+        fn next(&mut self) -> Option<Self::Item> {
+            while let Some(v) = self.iter.next() {
+                if let Some(v) = self.filter(v) {
+                    return Some(v)
+                }
+            }
+            None
```

Comment:
In the spirit of #517, it's possible to use internal iteration to replace these while-let loops with `find_map`!

Unfortunately, because rust lacks split borrows, it's **not** possible to _just_ write this:
```suggestion
            self.iter.find_map(|v| self.filter(v))
```

I've prototyped a work-around here: https://github.com/jswrenn/rust-itertools/blob/c7095488caca3db2dcc04b4233e20d57a7f730c9/src/duplicates_impl.rs

Reply:
Yeah I hit the same split borrow problem, but once again it looked simpler in my eyes to keep it with the white loop instead of introducing yet another struct just to make a function call work.

In any case, I just pushed your suggestion but it's failing the MSRV tests, do you know why? It's working on my local linux machine and the linux tests on the CI seem to pass :/

Reply:
Ok, I added a couple more trait bounds and it's working now. I don't know why it failed before